[bugfix] connect to ndisplay change and set node.order

psobolewskiPhD opened the pull request:
Closes: https://github.com/napari-threedee/napari-threedee/issues/183

In this PR I connect the dims.ndisplay event to `_on_ndisplay_changed` in base_manipulator init.
Then, in `_on_ndisplay_changed` I change the node.order so that the manipulator is properly restored when toggling back to 3D.

This fixes the issue.

Question:
There is another unconnected callback here:
https://github.com/napari-threedee/napari-threedee/blob/fdfac7246d18bb72dc5ec3c0989a4c6b252b14a7/src/napari_threedee/_backend/manipulator/napari_manipulator_backend.py#L202-L206

Would the node.order be better there? Should I connect that one instead?

Or should that be removed? The mouse/drag callbacks are handled in base_manipulator:
https://github.com/napari-threedee/napari-threedee/blob/fdfac7246d18bb72dc5ec3c0989a4c6b252b14a7/src/napari_threedee/manipulators/base_manipulator.py#L204-L220

(in base_manipulator on_ndisplay_change the manipulator is disabled in 2d and enabled in 3d)

Or should I connect the event in each manipulator type, under their `def _connect_events(self):`

Finally, in `base_manipulator`:
https://github.com/napari-threedee/napari-threedee/blob/fdfac7246d18bb72dc5ec3c0989a4c6b252b14a7/src/napari_threedee/manipulators/base_manipulator.py#L222-L224

The last line, the `remove`, it doesn't actually work. the list doesn't get modified.
I'm not sure how to actually remove the manipulator visual. Any advice?

alisterburt commented:
Amazing work! I'm not sure how to respond to your questions right now but what you've done looks good to me... I think having this in the base manipulator rather than each manipulator type is best

What do you think?

alisterburt commented:
No advice on removing the visual, I would probably just make it invisible rather than removing it...

psobolewskiPhD commented:
> I think having this in the base manipulator rather than each manipulator type is best

yeah, all the manipulators only work in 3D, so making it a global behavior makes sense.
I guess it's just whether to have it in BaseManipulator vs. NapariManipulatorBackend

kevinyamauchi commented:
> I guess it's just whether to have it in BaseManipulator vs. NapariManipulatorBackend

I didn't have a chance to look closely, but I think having it in either place is fine. Slight preference for in the `NapariManipulatorBackend` since the event, etc. will be napari specific. Thanks, @psobolewskiPhD !

psobolewskiPhD commented:
Thanks for the input @kevinyamauchi 
So if I am understanding, leave the existing BaseManipulator callback (enable/disable) do ndisplay_change
and add the order change to the NapariManipulatorBackend callback (which needs to actually be connected)?
Or do you mean to move the enable/disable stuff to the NapariManipulatorBackend ndisplay callback too?
It's basically the mouse callbacks plus visibility, so maybe it makes sense?

kevinyamauchi commented:
> So if I am understanding, leave the existing BaseManipulator callback (enable/disable) do ndisplay_change
and add the order change to the NapariManipulatorBackend callback (which needs to actually be connected)?

This one makes the most sense to me. That being said, I don't have a strong preference on this one, so if you do, please proceed as you see fit.

psobolewskiPhD commented:
I moved the node.order part to NapariManipulatorBackend.

I also think I have the proper `remove` concept now:
`self._backend.vispy_visual.parent = None`
based on the vispy docs, setting parent to none will remove the node from the scenograph:
https://vispy.org/api/vispy.scene.node.html#vispy.scene.node.Node.parent

And I added a test that fails on main. Using render plane manipulator because it was easy, but all the manipulators use NapariManipulatorBackend, so it should be sufficient?

kevinyamauchi commented:
Wow, nice work @psobolewskiPhD !

> And I added a test that fails on main. Using render plane manipulator because it was easy, but all the manipulators use NapariManipulatorBackend, so it should be sufficient?

Yes, I think this is fine.

Is this ready for review?

psobolewskiPhD commented:
I think so!
❤️

kevinyamauchi reviewed (approved):
This looks good to me! Thank you @psobolewskiPhD !
